### PR TITLE
feat(@astrojs/netlify): edge middleware support

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

This PR adds support for Netlify edge middleware. The edge middleware, in Netlify, is simply an edge function. The function will call the server origin - the page itself - and eventually return the response.

Here is some context and documentation: https://docs.netlify.com/edge-functions/api/#return-a-redirect

The code is very similar to the one implemented in Vercel. It was slightly adapted to Netlify.

## Testing

- new test cases added
- cut a preview release and see if I can use it in a pet project

## Docs

 /cc @withastro/maintainers-docs for feedback!

### CODE REVIEW

**Comment by ematipico** [comment]
!preview netlify-edge-middleware

File: packages/integrations/netlify/README.md [hunk 1]
```diff
@@ -97,6 +97,65 @@ Once you run `astro build` there will be a `dist/_redirects` file. Netlify will
...
+
+- must export a **default** function;
+- must **return** an `object`;
+- accepts an object with a `request` and `context` as properties;
+- `request` is typed as [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request);
+- `context` is typed as [`Context`](https://docs.netlify.com/edge-functions/api/#edge-function-types);
+
+#### Limitations and constraints
+
+When you opt in to this feature, there are few constraints to note:
```
Comment (Yan-Thomas) [inline packages/integrations/netlify/README.md:152]: two typos
```suggestion
When you opt-in to this feature, there are a few constraints to note:
```

File: packages/integrations/netlify/README.md [hunk 2]
```diff
@@ -97,6 +97,65 @@ Once you run `astro build` there will be a `dist/_redirects` file. Netlify will
...
+
+- The Edge middleware will always be the **first** function to receive the `Request` and the last function to receive `Response`. This an architectural constraint that follows the [boundaries set by Netlify](https://vercel.com/docs/concepts/functions/edge-middleware).
```
Comment (Yan-Thomas) [inline packages/integrations/netlify/README.md:154]: small typo
```suggestion
- The Edge middleware will always be the **first** function to receive the `Request` and the last function to receive `Response`. This is an architectural constraint that follows the [boundaries set by Netlify](https://vercel.com/docs/concepts/functions/edge-middleware).
```

**Review by Yan-Thomas — COMMENTED** [review]
Nice work @ematipico! Added a few suggestions for you 🙌

File: packages/integrations/netlify/README.md [hunk 3]
```diff
@@ -115,6 +115,65 @@ Once you run `astro build` there will be a `dist/_redirects` file. Netlify will
 > **Note**
 > You can still include a `public/_redirects` file for manual redirects. Any redirects you specify in the redirects config are appended to the end of your own.
 
+
+### Edge Middleware with Astro middleware
+
+The `@astrojs/netlify/functoins` adapter can automatically create an edge function that will act as "Edge Middleware", from an Astro middleware in your code base.
```
Comment (bluwy) [inline packages/integrations/netlify/README.md:121]:
```suggestion
The `@astrojs/netlify/functions` adapter can automatically create an edge function that will act as "Edge Middleware", from an Astro middleware in your code base.
```

**Comment by ematipico** [comment]
!preview netlify-edge-middleware

File: .changeset/great-days-judge.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
+'@astrojs/netlify': minor
+---
+
+When a project uses the new option Astro `build.excludeMiddleware`, the
+`@astrojs/vercel/functions` adapter will automatically create an Edge Middleware
+that will automatically communicate with the Astro Middleware.
+
+Check the [documentation](https://github.com/withastro/astro/blob/main/packages/integrations/netlify/README.md#edge-middleware-with-astro-middleware) for more details.
```
Comment (bluwy) [inline .changeset/great-days-judge.md:9]:
```suggestion
Check the [documentation](https://github.com/withastro/astro/blob/main/packages/integrations/vercel/README.md#edge-middleware-with-astro-middleware) for more details.
```

File: packages/integrations/netlify/src/shared.ts [hunk 5]
```diff
@@ -21,3 +52,65 @@ export async function createRedirects(
...
+
+	const baseDir = new URL('./.netlify/edge-functions/', dir);
+	await fs.promises.mkdir(baseDir, { recursive: true });
+
+	const manifestURL = new URL('./manifest.json', baseDir);
+	const _manifest = JSON.stringify(manifest, null, '  ');
+	await fs.promises.writeFile(manifestURL, _manifest, 'utf-8');
+}
+
+export async function copyFilesToEdgeDirectory() {}
```
Comment (bluwy) [inline packages/integrations/netlify/src/shared.ts:88]: Is this function used?

File: packages/integrations/netlify/test/edge-functions/prerender.test.ts [hunk 6]
```diff
@@ -71,6 +81,6 @@ Deno.test({
 			} catch (e) {}
 			assertEquals(prerenderedRouteExists, false);
 		});
-		await fixture.cleanup();
+		// await fixture.cleanup();
```
Comment (bluwy) [inline packages/integrations/netlify/test/edge-functions/prerender.test.ts:84]: Do we still need the cleanup here?

File: packages/integrations/netlify/test/edge-functions/edge-basic.test.ts [hunk 7]
```diff
@@ -3,30 +3,38 @@ import { assertEquals, assert, DOMParser } from './deps.ts';
...
+			await fixture.runBuild();
+		});
+		await t.step('Should correctly render the response', async () => {
+			// TODO: debug why build cannot be found in "await import"
+			// Seems that the files are created, but can't be read at this stage.
+			// Remove the cleanup at the end and run twice, and the test will pass.
+			// This is weird...
+			const { default: handler } = await import(
+				'./fixtures/edge-basic/.netlify/edge-functions/entry.js'
+			);
```
Comment (bluwy) [inline packages/integrations/netlify/test/edge-functions/edge-basic.test.ts:22]: I noticed the test is now unskipped and CI is passing. Is the TODO here still needed?

I remember digging into the Deno tests here and found the same weird issues.

**Comment by ematipico** [comment]
!preview netlify-edge-middleware

**Comment by ematipico** [comment]
!preview netlify-edge-middleware

**Comment by ematipico** [comment]
I am going to close the PR. Unfortunately, supporting "edge middleware" is impossible using the same trick we used in Vercel. From an edge function, it seems it's not possible to call the origin server (the actual lambda that contains the `entry` file).

It's a pity. Hopefully, things will change in the future.

**Comment by ematipico** [comment]
!preview netlify-edge-middleware

**Comment by ematipico** [comment]
!preview netlify-edge-middleware

File: packages/integrations/netlify/test/edge-functions/edge-basic.test.ts (hunk 7 above)
Comment (ematipico) [reply packages/integrations/netlify/test/edge-functions/edge-basic.test.ts]: I reverted the test command changes. I have no idea why Deno is not able to find the module, even it's there before the import execution, and the paths are correct.

**Review by matthewp — APPROVED** [review]

File: packages/integrations/netlify/src/middleware.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	astroMiddlewareEntryPointPath: URL,
+	outPath: string,
+	netlifyEdgeMiddlewareHandlerPath: URL
+): Promise<URL> {
+	const entryPointPathURLAsString = JSON.stringify(
+		fileURLToPath(astroMiddlewareEntryPointPath).replace(/\\/g, '/')
+	);
+
+	const code = edgeMiddlewareTemplate(entryPointPathURLAsString, netlifyEdgeMiddlewareHandlerPath);
+	const bundledFilePath = join(outPath, 'edgeMiddleware.js');
```
Comment (ematipico) [inline packages/integrations/netlify/src/middleware.ts:21]: Netlify wants a `.js` file. `.mjs` is not picked up.

File: packages/integrations/netlify/src/middleware.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+export default async function middleware(request, context) {
+	const url = new URL(request.url);
+	const ctx = createContext({ 
+		request,
+		params: {}
+	});
+	ctx.locals = ${handlerTemplateCall};
+	const next = async () => {	
+		request.headers.set(${JSON.stringify(ASTRO_LOCALS_HEADER)}, trySerializeLocals(ctx.locals));
+		return await context.next();
```
Comment (ematipico) [inline packages/integrations/netlify/src/middleware.ts:65]: We inject the locals in the headers, and when calling `next`, we will eventually call the origin server (the entry file) and get the response back.

**Comment by ematipico** [comment]
@matthewp and @bluwy the logic is now completed. Could you review again please?

File: .changeset/great-days-judge.md [hunk 10]
```diff
@@ -0,0 +1,10 @@
+---
+'@astrojs/netlify': minor
+---
+
+When a project uses the new option Astro `build.excludeMiddleware`, the
+`@astrojs/netlify/functions` adapter will automatically create an Edge Middleware
+that will automatically communicate with the Astro Middleware.
+
+Check the [documentation](https://github.com/withastro/astro/blob/main/packages/integrations/vercel/README.md#edge-middleware-with-astro-middleware) for more details.
```
Comment (bluwy) [inline .changeset/great-days-judge.md:9]:
```suggestion
Check the [documentation](https://github.com/withastro/astro/blob/main/packages/integrations/netlify/README.md#edge-middleware-with-astro-middleware) for more details.
```

My bad. For some reason I suggested changing to "vercel" before. It should be "netlify"

File: packages/integrations/netlify/README.md [hunk 11]
```diff
@@ -115,6 +115,65 @@ Once you run `astro build` there will be a `dist/_redirects` file. Netlify will
...
+
+Typings require the [`https://edge.netlify.com`](https://docs.netlify.com/edge-functions/api/#reference) types.
+
+> Netlify edge functions run in a Deno environment, so you would need to import types using URLs.
+> 
+> You can find more in the [Netlify documentation page](https://docs.netlify.com/edge-functions/api/#runtime-environment)
+
+```ts
+// src/netlify-edge-middleware.ts
+import type { Context } from "https://edge.netlify.com";
```
Comment (bluwy) [inline packages/integrations/netlify/README.md:148]: This is wild syntax 😄

**Pull request merged**